### PR TITLE
TST: sparse.linalg.LinearOperator: test lazy backends

### PR DESCRIPTION
**Pull request by lucascolley**
towards gh-23315

### CODE REVIEW

File: scipy/sparse/linalg/tests/test_interface.py [hunk 1]
```diff
@@ -14,16 +14,26 @@
...
 
+pytestmark = make_xp_pytest_marks(
+    (LinearOperator, "__init__"),
+    (LinearOperator, "matvec"),
+    (LinearOperator, "matmat"),
+    (LinearOperator, "dot"),
+    (LinearOperator, "rdot"),
+    (LinearOperator, "rmatvec"),
+    (LinearOperator, "rmatmat"),
+)
```
Comment (lucascolley) [inline scipy/sparse/linalg/tests/test_interface.py:36]: I think pretty much every test will call one of these. Maybe we also want `aslinearoperator`

File: scipy/sparse/linalg/tests/test_interface.py [hunk 2]
```diff
@@ -144,11 +153,18 @@ def get_matvecs(A):
             z = A+A
             assert len(z.args) == 2 and z.args[0] is A and z.args[1] is A
             z = 2*A
-            assert len(z.args) == 2 and z.args[0] is A and z.args[1] == 2
+            assert len(z.args) == 2
+            if not is_lazy_array(xp.empty(0)):
+                assert z.args[0] is A and z.args[1] == 2
```
Comment (lucascolley) [inline scipy/sparse/linalg/tests/test_interface.py:159]: the monkey-patching breaks these `is` checks

File: scipy/sparse/linalg/tests/test_interface.py [hunk 3]
```diff
@@ -144,11 +153,18 @@ def get_matvecs(A):
...
 
             array_object = type(xp.empty(0))
             assert isinstance(A.matvec(_asarray([1, 2, 3])), array_object)
-            with pytest.warns(FutureWarning, match="column vectors"):
+            ctx = (
+                pytest.warns(FutureWarning, match="column vectors")
+                if not is_jax_array(xp.empty(0))
+                else contextlib.nullcontext()
+            )
+            with ctx:
```
Comment (lucascolley) [inline scipy/sparse/linalg/tests/test_interface.py:168]: Dask has a concrete `shape` here so hits the warning path, JAX has an abstract shape so doesn't

File: scipy/sparse/linalg/tests/test_interface.py [hunk 4]
```diff
@@ -983,31 +993,29 @@ def matvec(v):
...
 @pytest.mark.parametrize("test_dtype", ALLDTYPES)
-@make_xp_test_case(interface.LinearOperator)
 def test_determine_lo_dtype_from_matvec(test_dtype, xp):
     if "longdouble" in test_dtype.__name__ and not is_numpy(xp):
         pytest.skip("longdoubles are only tested for `np`")
     # gh-19209
     scalar = xp.asarray(np.array(1, dtype=test_dtype))
     def mv(v):
-        return xp.asarray([scalar * v[0], v[1]])
+        return xp.stack([scalar * v[0], v[1]])
```
Comment (lucascolley) [inline scipy/sparse/linalg/tests/test_interface.py:1005]: this was just wrong, the `try`-`except` below was hiding the failure with array-api-strict

File: scipy/sparse/linalg/tests/test_interface.py [hunk 5]
```diff
@@ -983,31 +993,29 @@ def matvec(v):
...
     # - unless that fails (e.g. via overflow or no mixed exact-inexact promotion),
     #   in which case use the default integral dtype
-    try:
-        xp.asarray(2) + xp.asarray(2.0)
-    except TypeError:
-        expected = xpx.default_dtype(xp, kind="integral")
-    else:
-        expected = scalar.dtype
+    expected = scalar.dtype
+    if test_dtype in globals()["INEXACTDTYPES"]:
```
Comment (lucascolley) [inline scipy/sparse/linalg/tests/test_interface.py]: oops we can just use https://data-apis.org/array-api/draft/API_specification/generated/array_api.isdtype.html#isdtype

File: scipy/sparse/linalg/tests/test_interface.py [hunk 6]
```diff
@@ -1110,7 +1114,7 @@ def test_sparse_matmat_exception():
     xp_assert_equal((A @ B).matvec(np.ones(2)), np.ones(2))
 
 
-@make_xp_test_case(interface.LinearOperator)
+@pytest.mark.skip_xp_backends(np_only=True) # remaining reference with JAX
```
Comment (lucascolley) [inline scipy/sparse/linalg/tests/test_interface.py]:
```
ref        = <weakref at 0x17d307510; to 'scipy.sparse.linalg._interface.MatrixLinearOperator' at 0x17d3e79d0>
```

**Comment by lucascolley** [comment]
not sure what the most appropriate tol bump is for these JAX GPU failures

**Comment by steppi** [comment]
> not sure what the most appropriate tol bump is for these JAX GPU failures

I'm looking into it. The results look bad enough that it seems like we may want to `xfail` that test with JAX on GPU and possibly submit a JAX issue if we can find the core problem and there isn't a relevant issue already.

**Comment by steppi** [comment]
@lucascolley, I've found that the issue with JAX here is that it defaults to TensorFloat32 precision for matrix multiplication of float32 arrays when using an NVIDIA card that supports TensorFloat32. See https://github.com/jax-ml/jax/issues/4873. TensorFloat32 uses only half precision in the mantissa, so you can't expect much accuracy.

I suggest adding something like this to relevant part of `scipy/conftest.py`

```python
        # Ensure JAX doesn't default to less accurate TensorFloat32 precision matmuls 
        # for float32 inputs on GPUs that support this support.
        jax.config.update("jax_default_matmul_precision", "float32")
```
and documenting that it's recommended to set `jax_default_matmul_precision` to `"float32"` when using SciPy with the JAX backend in the part of the array API documentation that discusses `JAX_ENABLE_FLOAT64`. 

I've also found that these failures can be slightly flaky locally because no `rng`  is passed to `generate_broadcastable_shapes` here:

https://github.com/scipy/scipy/blob/0a0a989d71f5fad44a62a1c378be80bb0ea02619/scipy/sparse/linalg/tests/test_interface.py#L386-L388

I'll leave it up to you if you want to do everything here, or just xfail those tests in this PR and do the default_matmul changes in a follow-up. If you do everything here, then we shouldn't squash merge, but instead keep the commits separate.

File: scipy/sparse/linalg/tests/test_interface.py (hunk 1 above)
Comment (steppi) [reply scipy/sparse/linalg/tests/test_interface.py:36]: Yeah, I think it's fine to group these all up like this, especially since the capabilities are identical for all of them.

Note that `aslinearoperator` isn't going to be able to support the JIT even in principle. Usually `__init__` cannot either, but because `__init__` has no array args for `LinearOperator`, `jax_autojit` from `array_api_extra` will treat all args as static. I'm not sure why `aslinearoperator` doesn't have an `xp_capabilities` entry yet, but when it gets one it will need `jax_jit=False`.

**Comment by lucascolley** [comment]
> documenting that it's recommended to set `jax_default_matmul_precision` to `"float32"` when using SciPy with the JAX backend in the part of the array API documentation that discusses `JAX_ENABLE_FLOAT64`.

I can't find a docs section that mentions `JAX_ENABLE_FLOAT64` — would you be willing to open a PR that I could review?

File: scipy/sparse/linalg/tests/test_interface.py (hunk 2 above)
Comment (steppi) [reply scipy/sparse/linalg/tests/test_interface.py:159]: I think this could use a comment.

**Comment by steppi** [comment]
> > documenting that it's recommended to set `jax_default_matmul_precision` to `"float32"` when using SciPy with the JAX backend in the part of the array API documentation that discusses `JAX_ENABLE_FLOAT64`.
> 
> I can't find a docs section that mentions `JAX_ENABLE_FLOAT64` — would you be willing to open a PR that I could review?

Sorry, it used to be here in an earlier draft of those docs

https://github.com/scipy/scipy/blob/0a0a989d71f5fad44a62a1c378be80bb0ea02619/doc/source/dev/api-dev/array_api.rst?plain=1#L344-L352

but got removed. Yeah, I'll submit a PR that also fixes the other failures that popped up.

File: scipy/sparse/linalg/tests/test_interface.py [hunk 7]
```diff
@@ -384,7 +402,7 @@ def check_matvec(
         *batch_shape, M, N = op.shape
 
         u_broadcast, v_broadcast = generate_broadcastable_shapes(
-            2, ndim=3, min=0, max=5
+            2, ndim=3, min=0, max=5, rng=63
```
Comment (steppi) [inline scipy/sparse/linalg/tests/test_interface.py:408]: Good. It's surprising this escaped attention for so long. I'm not sure if you observed this, but when I just plugged in `rng=rng`, I actually wasn't seeing the failures locally.

File: scipy/sparse/linalg/tests/test_interface.py (hunk 6 above)
Comment (steppi) [inline scipy/sparse/linalg/tests/test_interface.py]: Is this inherently NumPy only, or something that there is hope of getting working on other backends? 

From the docs

```rst
Tests which are inherently NumPy only should avoid the ``xp`` fixture
altogether rather than using it with an ``np_only=True`` skip marker.
```

**Review by steppi — COMMENTED** [review]
Thanks @lucascolley. This looks good. Just a few very minor comments and then I think this is good to merge once https://github.com/scipy/scipy/pull/24724 gets in and this is rebased.

File: scipy/sparse/linalg/tests/test_interface.py (hunk 6 above)
Comment (lucascolley) [reply scipy/sparse/linalg/tests/test_interface.py]: not inherently NumPy only, as far as I can tell it actually does work for all backends apart from JAX. It may be something to do with the monkeypatching? But at the same time, I was guessing that for now we only care about this test passing for NumPy.

I'm also happy to only skip for JAX if you prefer.

Comment (steppi) [reply scipy/sparse/linalg/tests/test_interface.py]: I'm fine with either skipping only with JAX or marking as `numpy_only=True` as long as there's some kind of comment or reason string explaining why.

**Review by steppi — APPROVED** [review]
Thanks @lucascolley. I think it's in good enough shape now. I'll merge after the checks pass and you give the go-ahead. I don't really care that much about the  little suggestions about comments, so optional if you want to add them.

File: scipy/sparse/linalg/tests/test_interface.py (hunk 1 above)
Comment (lucascolley) [reply scipy/sparse/linalg/tests/test_interface.py:36]: hmm interesting, thanks, indeed we should add capabilities to `aslinearoperator`. Let's do that in a separate PR, as it's not clear to me exactly how best to do that.

Is this one of those situations where we will want `aslinearoperator` wrapped by `lazy_xp_function` in minimal tests of the function itself, but otherwise have it unwrapped where it is used in testing other things?

Comment (steppi) [reply scipy/sparse/linalg/tests/test_interface.py:36]: I'm not sure actually. I guess we can just check to see what actually works. For `scipy.spatial.transform` I found class instantiation based on array args didn't work under the JIT but I still don't have a perfect grasp on exactly what works and what doesn't.

File: scipy/sparse/linalg/tests/test_interface.py (hunk 6 above)
Comment (lucascolley) [reply scipy/sparse/linalg/tests/test_interface.py]: skipped just for JAX and added reason

File: scipy/sparse/linalg/tests/test_interface.py (hunk 1 above)
Comment (lucascolley) [reply scipy/sparse/linalg/tests/test_interface.py:36]: gh-24729

**Comment by lucascolley** [comment]
> I'll merge after the checks pass and you give the go-ahead.

go-ahead given!
